[FSTORE-1820] Documentation for REST API model deployments (#504)
logicalclocks/logicalclocks.github.io: docs/user_guides/mlops/serving/rest-api.md

# Hopsworks Model Serving REST API

## Introduction

> Hopsworks provides model serving using [KServe](https://kserve.github.io/website/) as the deployment framework and [Istio]() as the ingress gateway to the Kubernetes cluster. 
> 
> This document explains how to interact with a deployed model endpoint via REST.

Reviewer: ```suggestion
Hopsworks provides model serving capabilities by leveraging [KServe](https://kserve.github.io/website/) as the model serving platform and [Istio](https://istio.io/) as the ingress gateway to the model deployments. 

This document explains how to interact with a model deployment via REST API.
```

## Base URL

> The deployed model is accessible through the Istio ingress gateway. The URL to access the model is provided on the deployment page inside the Hopsworks UI. 

Reviewer: ```suggestion
Deployed models are accessible through the Istio ingress gateway. The URL to interact with a model deployment is provided on the model deployment page in the Hopsworks UI. 
```

The URL follows this format:
```text

> http://<ISTIO_GATEWAY_IP>/v1/models/<DEPLOYMENT_NAME>:predict

Reviewer: This is not 100% correct. The URL format depends on the model server. For example, /v1/models/<name>:predict works for python deployments, but not for TensorFlow or LLMs.

I would suggest something like: 
`http://<ISTIO_GATEWAY_IP>/<RESOURCE_PATH>`, where `RESOURCE_PATH` depends on the model server (e.g., vLLM, TensorFlow Serving, KServe sklearnserver).

=== "Python"
    ```python
    import requests

    data = {
        "inputs": [
            [
                4641025220953719,
                4920355418495856
            ]
        ]
    }

    headers = {
        "Host": "fraud.test.hopsworks.ai",
        "Authorization": "ApiKey 8kDOlnRlJU4kiV1Y.RmFNJY3XKAUSqmJZ03kbUbXKMQSHveSBgMIGT84qrM5qXMjLib7hdlfGeg8fBQZp",
        "Content-Type": "application/json"
    }

    response = requests.post(
        "http://10.87.42.108/v1/models/fraud:predict",
        headers=headers,
        json=data
    )
    print(response.json())
    
    ```
=== "Curl"
    ```bash
    curl -X POST "http://10.87.42.108/v1/models/fraud:predict" \
          -H "Host: fraud.test.hopsworks.ai" \
          -H "Authorization: ApiKey 8kDOlnRlJU4kiV1Y.RmFNJY3XKAUSqmJZ03kbUbXKMQSHveSBgMIGT84qrM5qXMjLib7hdlfGeg8fBQZp" \
          -H "Content-Type: application/json" \
          -d '{
                "inputs": [
                  [
                    4641025220953719,
                    4920355418495856
                  ]
                ]
              }'
    ```

## Example Response

> The model returns predictions in a JSON object. You can find more information [here](https://kserve.github.io/website/docs/concepts/architecture/data-plane/v1-protocol#response-format).

Reviewer: The model server responses also depend on the model server implementation :)
The `{ "predictions": [] }` format applies to sklearn/xgboost deployments, but TensorFlow Serving or vLLM returns a different format that the one specified in the link.
